ksqldb-engine/src/main/java/io/confluent/ksql/engine/OrphanedTransientQueryCleaner.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import static java.util.Objects.requireNonNull;
+
+import io.confluent.ksql.exception.KafkaResponseGetFailedException;
+import io.confluent.ksql.services.KafkaTopicClient;
+import io.confluent.ksql.services.ServiceContext;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OrphanedTransientQueryCleaner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OrphanedTransientQueryCleaner.class);
+
+  private final QueryCleanupService cleanupService;
+
+  public OrphanedTransientQueryCleaner(final QueryCleanupService cleanupService) {
+    this.cleanupService = requireNonNull(cleanupService);
+  }
+
+  public void cleanupOrphanedInternalTopics(
+      final ServiceContext serviceContext,
+      final Set<String> queryApplicationIds
+  ) {
+    final KafkaTopicClient topicClient = serviceContext.getTopicClient();
+    final Set<String> topicNames;
+    try {
+      topicNames = topicClient.listTopicNames();
```

Comment:
In theory, it should process just one set of commands at a time during startup (it shouldn't be possible for there to be more than one unprocessed file).  When it does that, it should call this just once.  So it does this list call on node startup and uses that list of topics to compare with all of the application query ids in the commands file.  It would seem to be more expensive to say for every queryApplicationId, check the existence of a given topic.  Not only is that more rpcs, it also might check plenty of completed transient queries the cleaned up with no issue.  I considered adding a "done cleaning" entry to the commands so I could find the set of only orphaned queries, but that seems potentially error prone as well and requires writing another entry.  This seemed a little simpler.

Tell me if you'd prefer the "done cleaning" approach.

Reply:
Yeah I agree the existing PR approach is better. At first, I didn't realize it happens just once on startup but since that's the case this is easier and less error-prone.